### PR TITLE
Restricting how we can end key searches

### PR DESCRIPTION
**Pull request by lemire**
We have an issue where when we search for a key and it is not found, we can end up in a poor state from which follow-up queries will cause trouble. See discussion https://github.com/simdjson/simdjson/discussions/1569

This is closely related to bug https://github.com/simdjson/simdjson/issues/1521

I had fixed that bug by allowing us to sometimes end the search while pointing at an arbitrary key within the object, see https://github.com/simdjson/simdjson/pull/1530 But that fix was not sufficient.

I tried harder to make it work in PR https://github.com/simdjson/simdjson/pull/1573 but the complexity seems to only grow.

The new PR goes back to an earlier idea expressed by PR https://github.com/simdjson/simdjson/pull/1524

The idea, simply put, is that instead of allowing us to finish the call to find_field_unordered_raw in any arbitrary state, we restrict what is possible:

1. If the key is found, then you are pointing at the value (right after checking :).
2. If the value is not found, then you are either outside of the object, or you are pointing at at last value accessed.

This more restrictive approach seems better at containing the complexity of the problem.

### CODE REVIEW

File: include/simdjson/generic/ondemand/value_iterator-inl.h [hunk 1]
```diff
@@ -86,7 +85,7 @@ simdjson_warn_unused simdjson_really_inline simdjson_result<bool> value_iterator
     // this object iterator will blithely scan that object for fields.
     if (_json_iter->depth() < depth() - 1) { return OUT_OF_ORDER_ITERATION; }
 #endif
-    has_value = false;
+    return false;
```
Comment (jkeiser) [inline include/simdjson/generic/ondemand/value_iterator-inl.h:88]: This is probably performance neutral (and clearer), but just checking: this change doesn't have an actual effect, right? `while (has_value)` should do nothing, and then we return false at the end?

File: include/simdjson/generic/ondemand/value_iterator-inl.h [hunk 2]
```diff
@@ -236,31 +263,44 @@ simdjson_warn_unused simdjson_really_inline simdjson_result<bool> value_iterator
...
+    // The has_next_field() advances the pointer and check that either ',' or '}' is found.
+    // It returns true if ',' is found, false otherwise. If anything other than ',' or '}' is found,
+    // then we are in error and we abort.
     if ((error = has_next_field().get(has_value) )) { abandon(); return error; }
   }
+  // Performance note: it maybe wasteful to rewind to the beginning when there might be
+  // no other query following. Indeed, it would require reskipping the whole object.
+  // Instead, you can just stay where you are. If there is a new query, there is always time
+  // to rewind.
+  if(at_first) { return false; }
```
Comment (jkeiser) [inline include/simdjson/generic/ondemand/value_iterator-inl.h:284]: Good catch!

File: include/simdjson/generic/ondemand/value_iterator-inl.h [hunk 3]
```diff
@@ -273,16 +313,28 @@ simdjson_warn_unused simdjson_really_inline simdjson_result<bool> value_iterator
...
 
     // No match: skip the value and see if , or } is next
     logger::log_event(*this, "no match", key, -2);
+    // The call to skip_child is meant to skip over the value corresponding to the key.
+    // After skip_child(), we are right before the next comma (',') or the final brace ('}').
     SIMDJSON_TRY( skip_child() );
+    // If we reached the end of the key-value pair we started from, then we know
+    // that the key is not there so we return false. We are either right before
+    // the next comma or the final brace.
+    if(_json_iter->position() == search_start) { return false; }
```
Comment (jkeiser) [inline include/simdjson/generic/ondemand/value_iterator-inl.h:328]: Ah! This was the bug. :+1:

**Review by jkeiser — COMMENTED** [review]
Looks correct to me! And probably performance neutral.

File: include/simdjson/generic/ondemand/value_iterator-inl.h (hunk 1 above)
Comment (lemire) [reply include/simdjson/generic/ondemand/value_iterator-inl.h:88]: The next bit is `while(has_value){` which, indeed, does not get executed if set `has_value = false` but that loop end mid-course in the function and then after that you have the rest of the function where you reenter the object and we try to restart the iteration... that part gets executed even if `has_value = false` is set.

**Comment by lemire** [comment]
@jkeiser Thanks. Merging.

**Pull request merged**